### PR TITLE
transformation: (convert_onnx_to_linalg) add lowering from onnx.Add to linalg.add

### LINKED ISSUE #2119: add lowering from onnx.Add to linalg.add
Blocked by #2116

### PR DESCRIPTION
**Pull request by kayode-gif**
resolves #2119

### CODE REVIEW

File: tests/filecheck/transforms/convert_onnx_to_linalg.mlir
Comment (superlopuh) [inline tests/filecheck/transforms/convert_onnx_to_linalg.mlir:1]: this doesn't seem to exercise the lowering

File: tests/filecheck/transforms/convert_onnx_to_linalg.mlir [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// RUN: xdsl-opt -p convert-onnx-to-linalg %s | filecheck %s
+
+
+func.func @main_graph(%0 : tensor<3x2xf32>, %1 : tensor<3x2xf32>) -> tensor<3x2xf32> {
```
Comment (superlopuh) [inline tests/filecheck/transforms/convert_onnx_to_linalg.mlir:4]: I would recommend not using func.func in this test, replacing it with test.op

**Review by superlopuh — COMMENTED** [review]
This looks like a great start, although it feels like we might as well start on the logic to convert between broadcasting rules in ONNX and MLIR/linalg. Namely, it looks like ONNX denotes broadcastable dimensions with `1`, whereas MLIR does with `?`. One approach would be to check whether any of the dimensions of the tensors being added are `1`, and raise `NotImplementedError`, and tackle this in a follow-up PR. The other is to add the helper in this PR directly. I'd maybe lean towards the first option, but the lowering itself is so tiiny that I'd be happy to do it all in one go. Your call.

File: tests/filecheck/transforms/convert_onnx_to_linalg.mlir
Comment (superlopuh) [reply tests/filecheck/transforms/convert_onnx_to_linalg.mlir:1]: yep exactly

**Review by superlopuh — APPROVED** [review]

**Comment by superlopuh** [comment]
Lowering looks great! I'm still concerned about broadcasting. Could you please add a check to the lowering to make sure that we raise `NotImplementedError` when the shape is dynamic?

**Comment by superlopuh** [comment]
And also file an issue to handle dynamic shapes in the ONNX to linalg conversion

**Comment by kayode-gif** [comment]
> And also file an issue to handle dynamic shapes in the ONNX to linalg conversion

File: xdsl/transforms/convert_onnx_to_linalg.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+@dataclass
+class AddOpLowering(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, add: onnx.Add, rewriter: PatternRewriter, /):
+        lhs_shape = add.lhs.type.get_shape()
```
Comment (superlopuh) [inline xdsl/transforms/convert_onnx_to_linalg.py:20]: just needs an isinstance check

**Review by superlopuh — APPROVED** [review]
🚀

**Review by compor — APPROVED** [review]
Beautiful, thank you!

**Pull request merged**